src/MongoDB.Driver.Core/Core/Authentication/AwsSignatureVersion4.cs

```diff
@@ -0,0 +1,160 @@
+                nonce: salt);
+
+            var canonicalHeaders = GetCanonicalHeaders(requestHeaders);
+            var signedHeaders = GetSignedHeaders(requestHeaders);
+
+            var canonicalRequest = string.Join("\n", "POST", "/", "", canonicalHeaders, "", signedHeaders, Hash(body));
+            var algorithm = "AWS4-HMAC-SHA256";
+            var credentialScope = $"{datestamp}/{region}/{service}/aws4_request";
+            var stringToSign = string.Join("\n", algorithm, timestamp, credentialScope, Hash(canonicalRequest));
+            var signature = GetSignature(stringToSign, secretKey, datestamp, region, service);
+
+            authorizationHeader = $"{algorithm} Credential={accessKey}/{credentialScope}, SignedHeaders={signedHeaders}, Signature={signature}";
+        }
+
+        //private static methods
+        private static string GetCanonicalHeaders(SortedDictionary<string, string> requestHeaders)
+        {
+            return string.Join("\n", requestHeaders.Select(x => $"{x.Key.ToLowerInvariant()}:{x.Value}"));
+        }
+
+        private static string GetRegion(string host)
+        {
+            if (host == "sts.amazonaws.com")
+            {
+                return "us-east-1";
+            }
+
+            var split = host.Split('.');
+            if (split.Count() > 1)
+            {
+                return split[1];
+            }
+
+            return "us-east-1";
+        }
+
+        private static SortedDictionary<string, string> GetRequestHeaders(
+            string body,
+            string contentType,
+            string host,
+            string timestamp,
+            string sessionToken,
+            byte[] nonce)
+        {
+            var requestHeaders = new SortedDictionary<string, string>
+            {
+                ["Content-Type"] = contentType,
+                ["Content-Length"] = body.Length.ToString(),
+                ["Host"] = host,
+                ["X-Amz-Date"] = timestamp,
+                ["X-MongoDB-GS2-CB-Flag"] = "n",
+                ["X-MongoDB-Server-Nonce"] = Convert.ToBase64String(nonce)
+            };
+            if (sessionToken != null)
+            {
+                requestHeaders["X-Amz-Security-Token"] = sessionToken;
+            }
+
+            return requestHeaders;
+        }
+
+        private static string GetSignature(string stringToSign, string secret, string date, string region, string service)
+        {
+            byte[] kDateBlock = Hmac256(Encoding.ASCII.GetBytes("AWS4" + secret), Encoding.ASCII.GetBytes(date));
+            byte[] kRegionBlock = Hmac256(kDateBlock, Encoding.ASCII.GetBytes(region));
+            byte[] kServiceBlock = Hmac256(kRegionBlock, Encoding.ASCII.GetBytes(service));
+            byte[] kSigningBlock = Hmac256(kServiceBlock, Encoding.ASCII.GetBytes("aws4_request"));
+
+            return BsonUtils.ToHexString(Hmac256(kSigningBlock, Encoding.ASCII.GetBytes(stringToSign)));
+        }
```

Comment:
Refactor `GetSignature` so that `secretAccessKey` is a `SecureString` and the `SecureString` is used as safely as possible:

    private static string GetSignature(string stringToSign, SecureString secretAccessKey, string date, string region, string service)
    {
        using (var decryptedSecureString = new DecryptedSecureString(secretAccessKey))
        {
            var aws4SecretAccessKeyChars = "AWS4".Concat(decryptedSecureString.GetChars()).ToArray();
            var aws4SecretAccessKeyBytes = Encoding.ASCII.GetBytes(aws4SecretAccessKeyChars);
            byte[] kDateBlock = Hmac256(aws4SecretAccessKeyBytes, Encoding.ASCII.GetBytes(date));
            Array.Clear(aws4SecretAccessKeyChars, 0, aws4SecretAccessKeyChars.Length);
            Array.Clear(aws4SecretAccessKeyBytes, 0, aws4SecretAccessKeyBytes.Length);
            byte[] kRegionBlock = Hmac256(kDateBlock, Encoding.ASCII.GetBytes(region));
            byte[] kServiceBlock = Hmac256(kRegionBlock, Encoding.ASCII.GetBytes(service));
            byte[] kSigningBlock = Hmac256(kServiceBlock, Encoding.ASCII.GetBytes("aws4_request"));

            return BsonUtils.ToHexString(Hmac256(kSigningBlock, Encoding.ASCII.GetBytes(stringToSign)));
        }
    }